Add retry_on/discard_on for better exception handling

dhh opened the pull request:
Declarative exception handling of the most common kind: retrying and discarding.

``` ruby
class RemoteServiceJob < ActiveJob::Base
  retry_on Net::OpenTimeout, wait: 30.seconds, attempts: 10

  def perform(*args)
    # Might raise Net::OpenTimeout when the remote service is down
  end
end

class SearchIndexingJob < ActiveJob::Base
  discard_on ActiveJob::DeserializationError

  def perform(record)
    # Will raise ActiveJob::DeserializationError if the record can't be deserialized
  end
end
```

Commit:
Add retry_on/discard_on for better exception handling

Commit:
Remove needless require

Commit:
Satisfy pedantic rubocop whitespace detection

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,76 @@
+      #
+      # ==== Options
+      # * <tt>:wait</tt> - Re-enqueues the job with the specified delay in seconds
+      # * <tt>:attempts</tt> - Re-enqueues the job the specified number of times
```
Might want to state the defaults here.

rafaelfranca commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,76 @@
+      #      # Might raise Net::OpenTimeout when the remote service is down
+      #    end
+      #  end
+      def retry_on(exception, wait: 3.seconds, attempts: 5)
```
I think we need to require `active_support/core_ext/numeric/time.rb` here.

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,76 @@
+      def retry_on(exception, wait: 3.seconds, attempts: 5)
+        rescue_from exception do |error|
+          logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
+          retry_job wait: wait if executions < attempts
```
Why pass on only `:wait` here? Since `retry_job` has an example of `rescue_from` that re-queues on another `queue`, maybe it's not that much of stretch that people might want to do that? I'd say the same goes for `priority`, though I can't see `wait_until` make sense as the time would be fixed.

Commit:
Mention defaults

Commit:
Require time extension for 3.seconds default

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,76 @@
+      # Discard the job with no attempts to retry, if the exception is raised. This is useful when the subject of the job,
+      # like an Active Record, is no longer available, and the job is thus no longer relevant.
+      #
+      # ==== Example
```
Nitpick: This says "Example" while the other docs say "Examples" (plural).

dhh replied:
If there's only 1 example, it doesn't make sense to me to pluralize.

kaspth replied:
The other cases only have one example as well.

dhh replied:
I'd rather fix those, then. Use plural when there are multiple examples and singular when just one.

kaspth replied:
That's what I meant to say, just seems I forgot to actually put the words down 👍

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,78 @@
+      #    discard_on ActiveJob::DeserializationError
+      #
+      #    def perform(record)
+      #      # Will raise ActiveJob::DeserializationError if the record can't be deserialized
```
I think our practice is to add a period to comments. Same goes for the other examples.

dhh replied:
I haven't been using that form consistently. Usually only use periods if there are multiple sentences.

Commit:
Allow retries to happen with different priority and queue

rafaelfranca commented:
:shipit:

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,80 @@
+      #      # Might raise Net::OpenTimeout when the remote service is down
+      #    end
+      #  end
+      def retry_on(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
```
Personally, I'd rearrange these arguments such that `wait`, `queue` and `priority` are together with `attempts` being last. That gives me the nice pleasure of seeing the same grouping as they have in the method body and brings them closer to their `retry_job` origin.

I'd do the same in the docs just above. 😁

dhh replied:
I can see that general point, but I think it's trumped by grouping default-value parameters together and option-value parameters together. wait/attempts are the overwhelmingly most likely to be used. Queue/priority much less so.

kaspth commented:
Luuuv the API, :shipit:

matthewd commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,80 @@
+      #  end
+      def retry_on(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
+        rescue_from exception do |error|
+          logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
```
This claims to be retrying when it should actually be saying it's giving up

matthewd commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,80 @@
+      def retry_on(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
+        rescue_from exception do |error|
+          logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
+          retry_job wait: wait, queue: queue, priority: priority if executions < attempts
```
I hesitate to suggest adding functionality, but it seems potentially useful to yield when we run out of retries (if `block_given?`).

dhh replied:
What's the use case? Presumably if someone can't be retried, after successive attempts, it's going to need human intervention to determine what should happen next? (Like fixing the bug first)

matthewd replied:
Logging in an application- / usage-specific place (an error tracker, a DB table, ... generally, something more structured to retain the details of the job than trusting someone to extract it from the log file).

Even more generally, anything that you might put in a `rescue_from` block, which is "do this after trying once", might equally apply for "do this after trying 2/3 times"... try to gently acquire a lock a couple of times, but if that's not working, we'll go for a more aggressive blocking operation, perhaps.

matthewd commented:
Alternative API:

```ruby
discard_on Net::OpenTimeout, retries: 10, wait: 30.seconds
```

```ruby
rescue_from Net::OpenTimeout, retries: 10, wait: 30.seconds do |error|
  # We failed to connect ten times; give up and send someone an email or something
end
```

Mostly, I like the fact this clearly states what happens when we run out of retries... `retry_on` feels a bit subtle for ".. and then almost-silently drop it on the floor".

Commit:
Proper logging when we bail on retrying after X attempts

dhh commented:
Ah, I see what you mean @matthewd. Actually what should happen after we try to retry a bunch of time and fail is that we should reraise and let the queue deal with it. Will fix that!

(Though I guess there's still an argument for ALSO allowing custom logic at that point, but).

matthewd commented:
Yeah, I guess I'm positing that (unlike "discard", say,) "retry" isn't a distinct error handling strategy, but an intrinsic.. step? attribute? of any overall error-handling plan. It just happens that the default behaviour is to make zero retries.

Commit:
Reraise instead of swallow exceptions that occur beyond the retry attempts

dhh commented:
Definitely. Failing to retry should absolutely not result in dropping the job. Just fixed that in the latest commit 👍

matthewd commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,85 @@
+            logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
+            retry_job wait: wait, queue: queue, priority: priority
+          else
+            logger.error "Stopped retrying #{self.class} due to a #{exception}, which reoccurred on #{executions} attempts. The original exception was #{error.cause.inspect}."
```
Super nitpick technicality: this particular exception may not have occurred during the previous executions. Maybe worth rephrasing to something like `"Stopping #{self.class} after #{executions} retries due to .."`... or maybe not worth bothering with.

dhh replied:
You mean because of inheritance? Or how would that particular exception not have been the cause?

matthewd replied:
If you have more than one `retry_on` (i.e., handling more than once possible exception) -- `executions` keeps counting up. The exception we're naming definitely caused *this* failure, but if we've done five attempts, they could've hit: `Resolv::ResolvTimeout` `Resolv::ResolvTimeout` `Resolv::ResolvTimeout` `Resolv::ResolvTimeout` `Net::OpenTimeout`. It's true we're giving up after 5 tries, but it's not strictly true to blame `Net::OpenTimeout` for reoccurring 5 times.

matthewd commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,85 @@
+      def retry_on(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
+        rescue_from exception do |error|
+          if executions < attempts
+            logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
```
Consider showing `error` instead of `exception`; the latter is more likely to be a vague parent class. "The original exception was nil." seems likely to be confusing, too.

This brushes with my "error reporting in general" project anyway, so possibly ignore for now.

dhh replied:
Not sure I follow? `exception` is the class, `error` is the object.

matthewd replied:
Because of inheritance, `error` could be an instance of a very specific exception class, while `exception` could be something so vague as `RuntimeError`.

matthewd commented:
> (Though I guess there's still an argument for ALSO allowing custom logic at that point, but).

FWIW, I think that's the thing I was arguing: if I write a custom `rescue_from`, I don't want to then have to implement my own retry mechanism -- that feels orthogonal to my decision on how I want to handle the "give up" step.

.. including on the new `discard_on` -- it seems right that the default after-retrying behaviour is to re-raise, but "try a few times then just forget it" seems just as likely as "try once then forget it".

robin850 commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+
+    # Reschedules the job to be re-executed. This is useful in combination
+    # with the +rescue_from+ option. When you rescue an exception from your job
```
I think you meant "the +rescue_from+ method" here, not option, no ?

dhh replied:
It's not meant as a programmatic option, but rather as "an option for dealing with exceptions". Other options include discard_on, retry_on.

Commit:
Executions counting is not a serialization concern

Let’s do it when we actually execute instead. Then the tests dealing
with comparable serializations won’t fail either!

Commit:
Use descriptive exception names

Commit:
Add exponentially_longer and custom wait algorithms

Commit:
Allow for custom handling of exceptions that persist beyond the retry attempts

dhh commented:
@matthewd Added the power to provide a custom handler if the retry attempts are unsuccessful.

dhh commented:
This is now ready to merge from my perspective, lest anyone has any last objections.

Commit:
Not needed

dhh commented:
Did not. Removed. Thanks.

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -0,0 +1,117 @@
+        case seconds_or_algorithm
+        when :exponentially_longer
+          (executions ** 4) + 2
+        when Integer
```
How does this fare with fixnums, bignums and rationals to name a few on Ruby 2.3?

Does it catch durations as well, if I passed in 1.hour?

dhh replied:
Integer is the parent class of those. So should be good. But no, doesn't catch duration. Could add that as a separate clause and .to_i it.

kaspth replied:
👍, since our documentation uses 3.seconds, I'd expect full duration support as a doc reader.

kaspth commented:
Left some questions, but otherwise good to me 😁

CodeClimate, on the other hand, seems to wish for some climate change 😋

Commit:
Please Rubocup

Commit:
Merge branch 'master' into retry-and-discard-jobs

kaspth commented on activejob/lib/active_job/exceptions.rb:
```diff
@@ -45,7 +45,7 @@ def retry_on(exception, wait: 3.seconds, attempts: 5, queue: nil, priority: nil)
           if executions < attempts
             logger.error "Retrying #{self.class} in #{wait} seconds, due to a #{exception}. The original exception was #{error.cause.inspect}."
             retry_job wait: determine_delay(wait), queue: queue, priority: priority
-          else          
```
I've never seen this much whitespace sacrificed for the glory of Rubocup! We must be saved now! Praise be our automaton savior 🙏🤖

Commit:
Fix tests against ActiveSupport::Durations

Commit:
Amend the CHANGELOG
